src/math/crypto/poseidon/poseidon.zig

```diff
@@ -0,0 +1,175 @@
+
+const COMPRESSED_ROUND_CONSTS = @import("./gen/constants.zig").POSEIDON_COMPRESSED_ROUND_CONSTS;
+const FULL_ROUNDS = @import("./gen/constants.zig").POSEIDON_FULL_ROUNDS;
+const PARTIAL_ROUNDS = @import("./gen/constants.zig").POSEIDON_PARTIAL_ROUNDS;
+
+fn mix(state: *[3]Felt252) void {
+    const t = Felt252.add(Felt252.add(state[0], state[1]), state[2]);
+
+    state[0] = Felt252.add(t, Felt252.mul(Felt252.two(), state[0]));
+    state[1] = Felt252.sub(t, Felt252.mul(Felt252.two(), state[1]));
+    state[2] = Felt252.sub(t, Felt252.mul(Felt252.three(), state[2]));
+}
+
+/// Linear layer for MDS matrix M = ((3,1,1), (1,-1,1), (1,1,2))
+/// Given state vector x, it returns Mx, optimized by precomputing t.
+fn round_comp(state: *[3]Felt252, idx: usize, full: bool) void {
+    if (full) {
+        inline for (0..3) |i| {
+            state[i] = Felt252.add(state[i], COMPRESSED_ROUND_CONSTS[idx + i]);
+            state[i] = Felt252.mul(Felt252.mul(state[i], state[i]), state[i]);
+        }
+    } else {
+        state[2] = Felt252.add(state[2], COMPRESSED_ROUND_CONSTS[idx]);
+        state[2] = Felt252.mul(Felt252.mul(state[2], state[2]), state[2]);
+    }
+    mix(state);
+}
+
+/// Poseidon permutation function.
+pub fn poseidon_permute_comp(state: *[3]Felt252) void {
+    var idx: usize = 0;
+
+    // Full rounds
+    for (0..(FULL_ROUNDS / 2)) |_| {
+        round_comp(state, idx, true);
+        idx += 3;
+    }
+
+    // Partial rounds
+    for (0..PARTIAL_ROUNDS) |_| {
+        round_comp(state, idx, false);
+        idx += 1;
+    }
+
+    // Full rounds
+    for (0..(FULL_ROUNDS / 2)) |_| {
+        round_comp(state, idx, true);
+        idx += 3;
+    }
+}
+
+/// Computes the Starknet Poseidon hash of x and y.
+pub fn poseidon_hash(x: Felt252, y: Felt252) Felt252 {
+    var state = [_]Felt252{ x, y, Felt252.two() };
+
+    poseidon_permute_comp(&state);
+
+    return state[0];
+}
+
+/// Computes the Starknet Poseidon hash of an arbitrary number of [Felt252]s.
+///
+/// Using this function is the same as using [PoseidonHasher].
+pub fn poseidon_hash_many(msgs: []const Felt252) Felt252 {
+    var state = [_]Felt252{Felt252.zero()} ** 3;
+
+    var i: usize = 0;
+    while (i + 1 < msgs.len) : (i += 2) {
+        state[0] = Felt252.add(state[0], msgs[i]);
+        state[1] = Felt252.add(state[1], msgs[i + 1]);
+
+        poseidon_permute_comp(&state);
+    }
+
+    const rem_len = msgs.len % 2;
+    if (rem_len == 1) {
+        state[0] = Felt252.add(state[0], msgs[msgs.len - 1]);
+    }
+
+    state[rem_len] = Felt252.add(Felt252.one(), state[rem_len]);
+
+    poseidon_permute_comp(&state);
+
+    return state[0];
+}
+
+/// Computes the Starknet Poseidon hash of a single [Felt252].
+pub fn poseidon_hash_single(x: Felt252) Felt252 {
+    var state = [_]Felt252{
+        x, Felt252.zero(), Felt252.one(),
+    };
+
+    poseidon_permute_comp(&state);
+
+    return state[0];
+}
+
+test "poseidon-hash" {
+    const test_data = [_][3]Felt252{
```

Comment:
can we include the comment from the ported source here? https://github.com/xJonathanLEI/starknet-rs/blob/0857bd6cd3bd34cbb06708f0a185757044171d8d/starknet-crypto/src/poseidon_hash.rs#L150